### ISSUE TITLE
Add some indication that Textpattern is doing something on long file operations

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#1697]
Steps to reproduce:

* Switch on 'prefers reduced motion' in the OS Accessibility settings.
* Upload a large image to Content>Images - a few MB - something that takes a while.
* Click to edit it.
* Drag/browse for another large image to Replace it.
* Click Upload... is Txp doing anything?

In Firefox, for example, the only indication is the static egg timer appearing in the tab bar. Especially with a crowded tab bar this is not very obvious. We could do with some UI-level element alongside the Upload button like we do when you click Save in the Write panel.

With reduced motion on, at least you see the static double-headed 'doing stuff' icon near the Save button, which is a nice visual clue that the button was clicked. With regular motion enabled, it animates. Could we do the same on the Images and Files panels when an action is taking place?

**Comment by petecooper** [comment]
![5269097](https://user-images.githubusercontent.com/226578/121254934-c2007880-c8a2-11eb-968e-986209fe53fe.gif)

**Comment by phiw13** [comment]
I thought Textpattern would display a progress bar, just as it does on the main Images panel. But upon testing, nope. Fun thing, the `<progress />` element is present in the source of the page (set to `display:none`) but it appears not to be used on the Edit panel.

In Safari at least, the browser progress bar is visible (OS standard widget in the  location bar) —and yeah, the Firefox indicator thingie  is beyond mysterious and hardly visible with multiple tabs open.

PS - Tested with `smd_thumbnail` installed, in case that changes something in the whole upload process of the main image

**Comment by Bloke** [comment]
Yes, using the progress bar on image/file upload in the Replace step would be ideal as it's consistent. Maybe it's supposed to use that but doesn't for some reason, and that's why there's a lack of other 'please wait' signal on those panels?

**Comment by petecooper** [comment]
Could this be subsumed by https://github.com/textpattern/textpattern/issues/1404 ?

**Comment by bloatware** [comment]
Please test [this](https://github.com/textpattern/textpattern/commit/7f1e30e236cc749a71ac64d66c5eab905bee7b5b) with a really large file/image.

**Comment by bloatware** [comment]
I'm slowly reworking the upload progress indicators and need your help. Now every file can have its own upload progressbar. I have added a voluntary dirty 'template' at the beginning of `textpattern.js` to see how it works.
``` js
textpattern.prefs.uploadPreview = '<div style="float:left;height:128px;width:128px;margin:5px;padding:5px;border:1px solid grey;overflow:hidden">'+
'<progress id="{hash}" class="txp-upload-progress hidden" style="width:128px;position:absolute" ></progress>{name}{preview}</div>';
```
It will be NOT part of core, but delegated to themes. I just need to know how to place the upload preview area inside the upload form in a way that it would be compatible both with batch uploads and individual file/image replace. @philwareham @phiw13 ?

**Comment by phiw13** [comment]
This will  certainly be appreciated by some users, I think. 

Your mark-up (a container child of the upload form) is OK, as far as I am concerned, proviso that the online styles are only part of the demo… (:-)). It is a bit odd-looking on the files panel (only a text string - the file name) unless the item is an image format.

How do you envision this to work? A base hook in the Core code, the theme would then attach  that template in some way? I remember that in the TXP 4.7 time frame, there was a possibility for an upload-preview for images, a theme just needed to provide the template as a `.json` file.

Is the fact that 2 progress bars are visible (in the preview-box and as part of the main UI) a feature?

PS a quick mock-up here: https://dev.l-c-n.com/txp/txp-content-image-preview.html

**Comment by bloatware** [comment]
Deal. As soon as the admin theme sets `textpattern.prefs.uploadPreview` pattern, core will leave it in its hands, including emptying `div.txp-upload-preview` container. Up to theme authors to decide whether they want to display previews, progress bars, % indicators or nothing at all on uploads.

**Comment by bloatware** [comment]
Every txp upload form now has a progress bar and a preview area configurable directly in admin themes or via `custom.css/custom.js`. Closing.

**Comment by Bloke** [comment]
Not tried the preview yet, but the progress bar is ace. Top work, sir!

**Comment by phiw13** [comment]
I’ve prepared an experimental Sandspace theme with optional upload previews.
https://l-c-n.com/file_download/5/phw_sandspace4.9.0-dev.zip

to enable the upload preview, add the following to your config.php: 

`define('upload_preview', true);`

Hopefully this approach is sane and relatively straightforward… (?)

–^–

Q: Does it actually make sense to have an upload preview panel on the images > edit and files > edit panels ? After all, the user never uploads more than one file at once (replace file/image, or upload thumbnail).

**Comment by bloatware** [comment]
You must be able to define (or not) `textpattern.prefs.uploadPreview` in `custom.js` depending on the values of `textpattern.event` and `textpattern.step`.

**Comment by phiw13** [comment]
It was (first) a question about whether having a preview for a single file was actually useful… Anyway, the question is kinda moot, as the user can (and probably does often) uploads a single file. Having a preview panel appear in a now-you-see-it-now-you-don’t way is not a great experience either. So everything will stay as it is for now.

Also, I’ll reorder the display – image first, then file title, and the progress widget on the individual files in the preview panel will be removed. Then I’ll leave at that as a proof of concept for this feature.

**Comment by phiw13** [comment]
PS - The progress bar on the replace image / file  works nicely, and certainly beats the invisible ‘activity widget’ used by the Firefox and Brave UI.
